### PR TITLE
KOGITO-2215 - generate operational and domain specific grafana dashboards

### PR DESCRIPTION
**Pull request by r00ta**
Jira link: https://issues.redhat.com/browse/KOGITO-2215

At the moment, when the `prometheus-addon-monitoring` addon is on, only one big dashboard is generated. 
For dmn endpoints it contains both operational and domain specific metrics graphs.
For rules endpoints it contains only operational metrics.

As requested in the jira issue, it's better to generate two separated dashboards: one for the operational metrics and one for the domain specific metrics. Some other pull requests will follow in the future, since we will add more metrics in both. 

With this PR, the grafana portal looks like this (using the kogito-app in `kogito-examples/dmn-drools-metrics-example`): 
Screenshot from 2020-05-19 15-19-36

The domain specific dashboard: 
Screenshot from 2020-05-19 15-19-48

And the operational one:
Screenshot from 2020-05-19 15-20-03

### CODE REVIEW

File: kogito-codegen/src/main/resources/grafana-dashboard-template/operational-dashboard-template.json [hunk 1]
```diff
@@ -481,7 +481,7 @@
     ]
   },
   "timezone": "",
-  "title": "$handlerName$ endpoint metrics",
+  "title": "",
```
Comment (evacchi) [inline kogito-codegen/src/main/resources/grafana-dashboard-template/operational-dashboard-template.json:484]: is this intended?

Comment (r00ta) [reply kogito-codegen/src/main/resources/grafana-dashboard-template/operational-dashboard-template.json:484]: Yeah, the title of the dashboard is set at code level when the template is imported

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java [hunk 2]
```diff
@@ -39,60 +41,52 @@ private GrafanaConfigurationWriter() {
...
-            logger.warn("Could not read the grafana dashboard template.", e);
-            return null;
-        }
+        JGrafana jgrafana = JGrafana.parse(template).setTitle(String.format("%s endpoint domain specific dashboard", endpoint));
 
         for (Decision decision : decisions) {
-            String type = decision.getVariable().getTypeRef().getLocalPart();
-            if (SupportedDecisionTypes.isSupported(type)) {
+            QName type = decision.getVariable().getTypeRef();
+            if (type != null && SupportedDecisionTypes.isSupported(type.getLocalPart())) {
```
Comment (danielezonca) [inline grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java:75]: Is it possible to have a `null` `type`?
Can you please log (warning?) the node ID in this case?

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java [hunk 3]
```diff
@@ -168,16 +171,29 @@ public DecisionContainerGenerator moduleGenerator() {
...
 
-        String dashboard = GrafanaConfigurationWriter.generateDashboardForDMNEndpoint(grafanaTemplatePath, resourceGenerator.getNameURL(), decisions);
+        String operationalDashboard = null;
+        String domainDashboard = null;
+
+        try {
+            operationalDashboard = GrafanaConfigurationWriter.generateOperationalDashboard(operationalDashboardDmnTemplate, resourceGenerator.getNameURL());
+            domainDashboard = GrafanaConfigurationWriter.generateDomainSpecificDMNDashboard(domainDashboardDmnTemplate, resourceGenerator.getNameURL(), decisions);
+        } catch (IOException e) {
+            logger.warn("Could not create the grafana dashboards for the endpoint " + resourceGenerator.getNameURL(), e);
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java:185]: I think `error` fit better

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java [hunk 4]
```diff
@@ -168,16 +171,29 @@ public DecisionContainerGenerator moduleGenerator() {
...
+        generatedFiles.add(new org.kie.kogito.codegen.GeneratedFile( org.kie.kogito.codegen.GeneratedFile.Type.RESOURCE,
+                                                                     "dashboards/operational-dashboard-" + resourceGenerator.getNameURL() + ".json",
+                                                                     operationalDashboard));
         generatedFiles.add(
                 new org.kie.kogito.codegen.GeneratedFile(
                         org.kie.kogito.codegen.GeneratedFile.Type.RESOURCE,
-                        "dashboards/dashboard-endpoint-" + resourceGenerator.getNameURL() + ".json",
-                        dashboard));
+                        "dashboards/domain-dashboard-" + resourceGenerator.getNameURL() + ".json",
+                        domainDashboard));
```
Comment (danielezonca) [inline kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java:196]: Can you please format both method invocation in the same way?

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java [hunk 5]
```diff
@@ -39,60 +41,52 @@ private GrafanaConfigurationWriter() {
...
      */
-    public static String generateDashboardForEndpoint(String templatePath, String handlerName) {
+    public static String generateOperationalDashboard(String templatePath, String handlerName) throws IOException {
         String template = readStandardDashboard(templatePath);
-        return customizeTemplate(template, handlerName);
+        template = customizeTemplate(template, handlerName);
+        JGrafana jgrafana;
+        jgrafana = JGrafana.parse(template).setTitle(String.format("%s endpoint operational dashboard", handlerName));
+
+        return jgrafana.serialize();
```
Comment (danielezonca) [inline grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java:56]: Can you please `try-catch` the checked exception, log it and propagate a runtime exception?

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java [hunk 6]
```diff
@@ -39,60 +41,52 @@ private GrafanaConfigurationWriter() {
...
             }
         }
 
-        try {
-            return jgrafana.serialize();
-        } catch (IOException e) {
-            logger.warn("Could not serialize the grafana dashboard template.", e);
-            return null;
-        }
+        return jgrafana.serialize();
```
Comment (danielezonca) [inline grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java:86]: Why have you decided to propagate `IOException`? I prefer the previous code :)

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java (hunk 2 above)
Comment (r00ta) [reply grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java:75]: I found during the demo that yes, this is possible for some dmn model. I will add a log

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java (hunk 3 above)
Comment (r00ta) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java:185]: Fine, do you think we should also stop the execution? For me if we log it as `error`, then we should not move forward with the compilation wdyt @danielezonca ?

Comment (r00ta) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java:185]: Will move to a runtime exception btw, so the execution will terminate

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java (hunk 4 above)
Comment (r00ta) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java]: Fixed

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java (hunk 5 above)
Comment (r00ta) [reply grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java]: Done, feel free to close resolve the conversation if you are ok with the new implementation

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java (hunk 6 above)
Comment (r00ta) [reply grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java]: Done

File: grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java (hunk 2 above)
Comment (danielezonca) [reply grafana-api/src/main/java/org/kie/kogito/grafana/GrafanaConfigurationWriter.java]: Nice to know :)

File: kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java (hunk 3 above)
Comment (danielezonca) [reply kogito-codegen/src/main/java/org/kie/kogito/codegen/decision/DecisionCodegen.java]: I think it is fine to stop the execution for now. If we want to make dashboard generation as optional, we will do in another ticket :)

**Comment by r00ta** [comment]
@jiripetrlik ping

**Pull request merged**